ai-quick-actions/ssh-notebook-session.md

```diff
@@ -0,0 +1,199 @@
+    - [Create Bastion Session](#create-bastion-session)
+    - [Connect to the Bastion](#connect-to-the-bastion)
+    - [Tunnel](#tunnel)
+    - [VSCode Configuration](#vscode-configuration)
+
+## Introduction
+Enabling the SSH tunnel to OCI Data Science notebooks provides the ability to do remote debugging and coding using tools like VSCode Remote-SSH or PyCharm's equivalent feature. Additionally, you can connect to the instance for manual debugging.
+
+## Prerequisite
+- Create a [Notebook Session](https://docs.oracle.com/en-us/iaas/data-science/using/manage-notebook-sessions.htm) in OCI Data Science Service using your own [VCN](https://docs.oracle.com/en-us/iaas/Content/Network/Tasks/VCNs.htm) and [Private Subnet](https://docs.oracle.com/en-us/iaas/Content/Network/Concepts/privateaccess.htm).
+
+## Step-by-Step Guide
+This guide outlines the steps required to enable SSH tunneling.
+
+### Reveal the Notebook Private IP
+Create a file named `nb_ip.py` and add the following Python code to reveal the private IP address of the notebook:
+
+```python
+import oci
+import os
+
+print(f"- get NB_SESSION_OCID")
+NB_OCID = os.environ.get("NB_SESSION_OCID")
+if NB_OCID is None:
+    raise Exception("NB_SESSION_OCID environment variable is required.")
+
+print(f"- init RP")
+RP = os.environ.get("OCI_RESOURCE_PRINCIPAL_VERSION", "UNDEFINED")
+
+if not RP or RP == "UNDEFINED":
+    # LOCAL RUN
+    config = oci.config.from_file("~/.oci/config", "DEFAULT")
+    dsc = oci.data_science.DataScienceClient(config=config)
+    network = oci.core.VirtualNetworkClient(config=config)
+else:
+    # NB RP
+    signer = oci.auth.signers.get_resource_principals_signer()
+    dsc = oci.data_science.DataScienceClient(config={}, signer=signer)
+    network = oci.core.VirtualNetworkClient(config={}, signer=signer)
+
+print(f"- get notebook data")
+nb = dsc.get_notebook_session(NB_OCID).data
+print(f"- notebook data: {nb}")
+
+if not nb.notebook_session_configuration_details or not nb.notebook_session_configuration_details.subnet_id:
+    raise Exception("IP Address cannot be determined when Default Networking is in use! For SSH tunneling to work, you must specify your own VCN and Private subnet!")
+
+print(f"- get notebook subnet")
+subnet_id = nb.notebook_session_configuration_details.subnet_id
+print(f"- notebook subnet ocid: {subnet_id}")
+
+print(f"- get subnet data")
+subnet = network.get_subnet(subnet_id).data
+
+print("- fetching subnet private IPs")
+private_ips = oci.pagination.list_call_get_all_results(
+    network.list_private_ips, subnet_id=subnet_id
+).data
+
+print("- finding the private IP for notebook session")
+display_name = NB_OCID.split(".")[-1]
+private_ip = [i for i in private_ips if i.display_name == display_name][0]
+
+print(f"- private IP address: {private_ip.ip_address}")
+```
+
+### Run the Script
+Run the script in the notebook to reveal the private IP address. Note the private IP address provided in the output:
+
+```bash
+python nb_ip.py
```

Comment:
why don't we put this into ADS in a model so it can be run with `python -m oracle_ads reveal_ip` similar to how you can run `python -m site` to reveal location of site-packages

Reply:
This is a good idea, we can definitely do this!